### PR TITLE
edit yaml funcs

### LINKED ISSUE #15: Improve yaml library

### PR DESCRIPTION
**Pull request by mrtkrkrt**
resolves #15 

Test will arrange.

### CODE REVIEW

File: .vscode/cmake-kits.json [hunk 1]
```diff
@@ -16,8 +16,8 @@
     {
       "name": "Qt Linux Kit",
       "compilers": {
-          "CC": "/usr/bin/gcc",
-          "CXX": "/usr/bin/g++"
+          "CC": "/usr/bin/gcc-8",
+          "CXX": "/usr/bin/g++-8"
```
Comment (rendayigit) [inline .vscode/cmake-kits.json:20]: This is not related to the current issue. Please revert this.

File: common.h [hunk 2]
```diff
@@ -9,10 +9,10 @@ using std::filesystem::current_path;
 
 namespace Path {
 const QString project_root = QString::fromStdString(current_path().string()) + "/../../";
-const QString paths_Yaml = project_root + "Setup/Paths.yaml";
-const QString bin_Dir = project_root + QString::fromStdString(Yaml::getText<std::vector<std::string>>(Yaml::getNodeByTag<std::vector<YAML::Node>>(paths_Yaml.toStdString(), "bin_Dir").at(0), "bin_Dir").at(0));
-const QString server_cmds_Yaml = project_root + QString::fromStdString(Yaml::getText<std::vector<std::string>>(Yaml::getNodeByTag<std::vector<YAML::Node>>(paths_Yaml.toStdString(), "server_cmds_Yaml").at(0), "server_cmds_Yaml").at(0));
-const QString client_cmds_Yaml = project_root + QString::fromStdString(Yaml::getText<std::vector<std::string>>(Yaml::getNodeByTag<std::vector<YAML::Node>>(paths_Yaml.toStdString(), "client_cmds_Yaml").at(0), "client_cmds_Yaml").at(0));
-const QString config_Yaml = project_root + QString::fromStdString(Yaml::getText<std::vector<std::string>>(Yaml::getNodeByTag<std::vector<YAML::Node>>(paths_Yaml.toStdString(), "config_Yaml").at(0), "config_Yaml").at(0));
+const QString paths_Yaml = "/home/pc/Desktop/SpaceLoader/Setup/Paths.yaml";
```
Comment (rendayigit) [inline common.h:12]: Please revert this line.

File: lib/YAML/yaml.cpp [hunk 3]
```diff
@@ -4,24 +4,50 @@ using namespace std;
...
     YAML::Node config = YAML::LoadFile(yamlFilePath);
     vector<YAML::Node> nodes = {};
 
-    vector<YAML::Node> result = Yaml::searchNodeByTag(config, tagName, tagValue, nodes);
-    return result;
+    return Yaml::searchNodeByTag(config, tagName, tagValue, nodes).at(0);
+}
+
+template <typename T>
+T Yaml::getMultipleNodeByTag(const string &yamlFilePath, const string &tagName,
```
Comment (rendayigit) [inline lib/YAML/yaml.cpp:16]: `getNodeVectorByTag` would be a better name.

File: lib/YAML/yaml.cpp [hunk 4]
```diff
@@ -4,24 +4,50 @@ using namespace std;
...
+                             const string &tagValue) {
+    YAML::Node config = YAML::LoadFile(yamlFilePath);
+    vector<YAML::Node> nodes = {};
+
+    return Yaml::searchNodeByTag(config, tagName, tagValue, nodes);
+}
+
+template <typename T>
+T Yaml::getSingleNodeByTag(const string &yamlFilePath, const string &tagName) {
```
Comment (rendayigit) [inline lib/YAML/yaml.cpp:25]: `getNodeByTag` is better.

File: lib/YAML/yaml.cpp [hunk 5]
```diff
@@ -4,24 +4,50 @@ using namespace std;
...
+    YAML::Node config = YAML::LoadFile(yamlFilePath);
+    vector<YAML::Node> nodes = {};
+
+    return Yaml::searchNodeByTag_(config, tagName, nodes).at(0);
 }
 
 template <typename T>
-T Yaml::getNodeByTag(const string &yamlFilePath, const string &tagName) {
+T Yaml::getMultipleNodeByTag(const string &yamlFilePath, const string &tagName) {
```
Comment (rendayigit) [inline lib/YAML/yaml.cpp:33]: `getNodeVectorByTag`  is better.

File: lib/YAML/yaml.cpp [hunk 6]
```diff
@@ -4,24 +4,50 @@ using namespace std;
...
     YAML::Node config = YAML::LoadFile(yamlFilePath);
     vector<YAML::Node> nodes = {};
 
     return Yaml::searchNodeByTag_(config, tagName, nodes);
 }
 
 template <typename T>
-T Yaml::getText(const YAML::Node &node, const string &tagName) {
+T Yaml::getSingleText(const YAML::Node &node, const string &tagName) {
```
Comment (rendayigit) [inline lib/YAML/yaml.cpp:41]: We should go with `getText`.

File: lib/YAML/yaml.cpp [hunk 7]
```diff
@@ -31,15 +57,32 @@ T Yaml::getText(const YAML::Node &node, const string &tagName) {
 }
 
 template <typename T>
-T Yaml::getText(const std::string &yamlFilePath, const std::string &tagName) {
+T Yaml::getSingleText(const std::string &yamlFilePath, const std::string &tagName) {
```
Comment (rendayigit) [inline lib/YAML/yaml.cpp:60]: `getText` is better.

File: lib/YAML/yaml.cpp [hunk 8]
```diff
@@ -31,15 +57,32 @@ T Yaml::getText(const YAML::Node &node, const string &tagName) {
...
+T Yaml::getMultipleText(const std::string &yamlFilePath, const std::string &tagName) {
     YAML::Node config = YAML::LoadFile(yamlFilePath);
     vector<string> texts = {};
 
     return Yaml::searchText(config, tagName, texts);
 }
 
 template <typename T>
-T Yaml::getNodeByPath(const string &yamlFilePath, const string &path) {
+T Yaml::getSingleNodeByPath(const string &yamlFilePath, const string &path) {
```
Comment (rendayigit) [inline lib/YAML/yaml.cpp:76]: `getNodeByPath` is better.

**Review by rendayigit — CHANGES_REQUESTED** [review]
My main request is for all these changes are to rename the ones that are named "multiple" to "Vector". (Like "getMultipleNodeByTag" to "getNodeVectorByTag"). And rename the ones that are "Single" to their previous names. (Like "getSingleText" to "getText"). Also please revert unrelated files like the files under .vscode dir. We may change them in a separate branch if necessary.

**Review by rendayigit — CHANGES_REQUESTED** [review]
Rename "Vectors" to "Lists". So it becomes "getNodeList" instead of "getNodeVector".
Also "getVectorText" is not right. "getTextList" is better.
And merge main to this branch.

**Comment by mrtkrkrt** [comment]
> Rename "Vectors" to "Lists". So it becomes "getNodeList" instead of "getNodeVector". Also "getVectorText" is not right. "getTextList" is better. And merge main to this branch.

Done.

File: lib/YAML/yaml.h [hunk 9]
```diff
@@ -11,33 +11,53 @@ class Yaml {
...
     static T getText(const std::string &yamlFilePath, const std::string &tagName);
 
+    template <typename T>
+    static T getTextList(const std::string &yamlFilePath, const std::string &tagName);
+
     template <typename T>
     static T getNodeByPath(const std::string &yamlFilePath, const std::string &path);
 
+    template <typename T>
+    static T getNodeListByPath(const std::string &yamlFilePath, const std::string &path);
```
Comment (rendayigit) [inline lib/YAML/yaml.h:40]: Also add getNodeByPath

**Review by rendayigit — CHANGES_REQUESTED** [review]

File: lib/YAML/yaml.h (hunk 9 above)
Comment (mrtkrkrt) [reply lib/YAML/yaml.h:40]: We have already defined the "getNodeByTag" function above the specified field. :)

File: tests/TCP/test_tcp.cpp [hunk 10]
```diff
@@ -1,4 +1,3 @@
-
 #include <QtConcurrent/QtConcurrent>
```
[the remark this answers is not in the document]
Comment (rendayigit) [reply tests/TCP/test_tcp.cpp:1]: Move these changes to a different branch if they are important.

File: tests/CMakeLists.txt [hunk 11]
```diff
@@ -18,15 +18,15 @@ find_package(Qt${QT_VERSION_MAJOR} COMPONENTS Core Quick REQUIRED) # if error ru
 find_package(Qt5Widgets REQUIRED)
```
Comment (rendayigit) [inline tests/CMakeLists.txt:18]: Ucomment the tests so we can merge to main.

**Review by rendayigit — CHANGES_REQUESTED** [review]

File: tests/TCP/test_tcp.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (mrtkrkrt) [reply tests/TCP/test_tcp.cpp:1]: Done.

File: tests/CMakeLists.txt (hunk 11 above)
Comment (mrtkrkrt) [reply tests/CMakeLists.txt]: Uncommented tests. For Now, only ThreadSafety test is not working.

**Comment by mrtkrkrt** [comment]
> Rename "Vectors" to "Lists". So it becomes "getNodeList" instead of "getNodeVector". Also "getVectorText" is not right. "getTextList" is better. And merge main to this branch.

Done

**Review by rendayigit — APPROVED** [review]

**Pull request merged**